x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -131,98 +127,93 @@ private void createFollowerIndex(
             return;
         }
 
-        ActionListener<Boolean> handler = ActionListener.wrap(
-                result -> {
-                    if (result) {
-                        initiateFollowing(request, listener);
-                    } else {
-                        listener.onResponse(new PutFollowAction.Response(true, false, false));
-                    }
-                },
-                listener::onFailure);
-        // Can't use create index api here, because then index templates can alter the mappings / settings.
-        // And index templates could introduce settings / mappings that are incompatible with the leader index.
-        clusterService.submitStateUpdateTask("create_following_index", new AckedClusterStateUpdateTask<Boolean>(request, handler) {
+        String remoteCluster = request.getRemoteCluster();
+        boolean waitForRestore = request.getWaitForCompletion();
 
-            @Override
-            protected Boolean newResponse(final boolean acknowledged) {
-                return acknowledged;
-            }
+        Client client = CcrLicenseChecker.wrapClient(this.client, threadPool.getThreadContext().getHeaders());
 
-            @Override
-            public ClusterState execute(final ClusterState currentState) throws Exception {
-                String followIndex = request.getFollowRequest().getFollowerIndex();
-                IndexMetaData currentIndex = currentState.metaData().index(followIndex);
-                if (currentIndex != null) {
-                    throw new ResourceAlreadyExistsException(currentIndex.getIndex());
-                }
+        ActionListener<PutFollowAction.Response> followingListener = waitForRestore ? listener : ActionListener.wrap(() -> {});
```

Comment:
I was a bit confused by the fact by the  `if waitForRestore == false` statement at the end of this method.
 
The `initiateRestore` method happen asynchronously on the snapshot threadpool, so if there is a failure in `RestoreService#restoreSnapshot` the user wouldn't be know about from the response.

So maybe we can rewrite this line here to this:

```java
final ActionListener<PutFollowAction.Response> followingListener;
if (request.getWaitForCompletion()) {
   followingListener = listener;
} else {
   followingListener = ActionListener.wrap(() -> {});
   listener.onResponse(new PutFollowAction.Response(false, false, false));
}
```

Then the wait for completion logic is in a single place.

Reply:
Also is there a dedicated test for `waitForCompletion` is set to `false`? Maybe I missed it.

Reply:
I added a test and made this changed with the listener.

Reply:
:+1: thanks!